### PR TITLE
Add static Squared_euclidean_distance, Shapes to face object and Jittering

### PR DESCRIPTION
**Pull request by Danile71**

### CODE REVIEW

**Comment by Danile71** [comment]
Jittering works now

Without jittering euclid distance- 0.13405496
With jittering = 10 - 0.114221856
With jittering = 100 - 0.105019964

File: face.go [hunk 1]
```diff
@@ -28,15 +29,16 @@ type Recognizer struct {
...
 	Descriptor Descriptor
 }
 
 // Descriptor holds 128-dimensional feature vector.
 type Descriptor [128]float32
 
 // New creates new face with the provided parameters.
-func New(r image.Rectangle, d Descriptor) Face {
-	return Face{r, d}
+func New(r image.Rectangle, s []image.Point, d Descriptor) Face {
```
Comment (Kagami) [inline face.go]: Would be better to make another function `NewWithShape`, to avoid breaking API

File: face.go [hunk 2]
```diff
@@ -28,15 +29,16 @@ type Recognizer struct {
 // Face holds coordinates and descriptor of the human face.
 type Face struct {
 	Rectangle  image.Rectangle
+	Shapes     []image.Point
```
Comment (Kagami) [inline face.go]: Better to put new fields at the end

File: face.go [hunk 3]
```diff
@@ -58,15 +60,25 @@ func NewRecognizer(modelDir string) (rec *Recognizer, err error) {
 	return
 }
 
-func (rec *Recognizer) recognize(imgData []byte, maxFaces int) (faces []Face, err error) {
+func (rec *Recognizer) Config(size int, padding float32, jittering int) {
```
Comment (Kagami) [inline face.go]: Would be better to pass config when creating recognizer, e.g. `NewRecognizerWithConfig`

File: face.go [hunk 4]
```diff
@@ -106,7 +118,14 @@ func (rec *Recognizer) recognize(imgData []byte, maxFaces int) (faces []Face, er
 	return
 }
 
-func (rec *Recognizer) recognizeFile(imgPath string, maxFaces int) (face []Face, err error) {
+func Squared_euclidean_distance(sample Descriptor, testSample Descriptor) (distance float32) {
```
Comment (Kagami) [inline face.go]: We can implement it in pure Go, see https://en.wikipedia.org/wiki/Euclidean_distance#Squared_Euclidean_distance

Comment (Kagami) [reply face.go]: Also should use CamelCase for naming

File: face.go [hunk 5]
```diff
@@ -115,21 +134,21 @@ func (rec *Recognizer) recognizeFile(imgPath string, maxFaces int) (face []Face,
...
+	return rec.recognize(type_, imgData, maxFaces)
 }
 
 // Recognize returns all faces found on the provided image, sorted from
 // left to right. Empty list is returned if there are no faces, error is
 // returned if there was some error while decoding/processing image.
 // Only JPEG format is currently supported. Thread-safe.
-func (rec *Recognizer) Recognize(imgData []byte) (faces []Face, err error) {
-	return rec.recognize(imgData, 0)
+func (rec *Recognizer) Recognize(type_ int, imgData []byte) (faces []Face, err error) {
```
Comment (Kagami) [inline face.go]: We could make new function e.g. `RecognizeCNN` to avoid breaking API

File: face.go [hunk 6]
```diff
@@ -115,21 +134,21 @@ func (rec *Recognizer) recognizeFile(imgPath string, maxFaces int) (face []Face,
...
+	return rec.recognize(type_, imgData, 0)
 }
 
 // RecognizeSingle returns face if it's the only face on the image or
 // nil otherwise. Only JPEG format is currently supported. Thread-safe.
-func (rec *Recognizer) RecognizeSingle(imgData []byte) (face *Face, err error) {
-	faces, err := rec.recognize(imgData, 1)
+func (rec *Recognizer) RecognizeSingle(type_ int, imgData []byte) (face *Face, err error) {
```
Comment (Kagami) [inline face.go]: Same as above regarding API

File: facerec.cc [hunk 7]
```diff
@@ -197,3 +260,27 @@ void facerec_free(facerec* rec) {
...
 }
+
+float squared_euclidean_distance(const float* c_sample, const float* c_test_sample) {
+	descriptor sample = mat(c_sample, DESCR_LEN, 1);
+	descriptor test_sample = mat(c_test_sample, DESCR_LEN, 1);
+	auto dist_func = dlib::squared_euclidean_distance();
+	return dist_func(sample, test_sample);
+}
+
+std::vector<matrix<rgb_pixel>> jitter_image(
```
Comment (Kagami) [inline facerec.cc]: Should be static

File: face.go (hunk 1 above)
Comment (Danile71) [reply face.go]: Fixed

File: face.go (hunk 2 above)
Comment (Danile71) [reply face.go]: Fixed

File: face.go (hunk 5 above)
Comment (Danile71) [reply face.go]: Fixed

File: facerec.cc (hunk 7 above)
Comment (Danile71) [reply facerec.cc]: Fixed

**Comment by Danile71** [comment]
Code formated

**Comment by Kagami** [comment]
Thank you, merged.

**Pull request merged**